### PR TITLE
Feature : allow players to switch between unselected investigators

### PR DESCRIPTION
**Pull request by metalalive**
Summary : 
- API endpoint `/games/{game_id}/investigator` , `PATCH` method.
- use-case layer : SwitchInvestigatorUseCase
- new function `switch_character()` in `Game` domain model

### CODE REVIEW

File: backend/app/domain/game.py [hunk 1]
```diff
@@ -75,6 +75,32 @@ def assign_character(self, investigator: Investigator) -> Optional[GameError]:
...
+    def _clear_character_selection(
+        self, investigator: Investigator
+    ) -> Optional[GameError]:
+        # TODO, figure out better design options
+        if investigator in self._investigators:
+            self._investigators[investigator] = False
+        else:
+            return GameError("invalid-investigator")
+
+    def switch_character(
```
Comment (teds-lin) [inline backend/app/domain/game.py:87]: 看起來有點亂，思考一下怎麼調整比較好
感覺應該不用這麼多error判斷

另外在我在研究這段code的時候，才注意到目前混用了有兩套error處理方式
1.   raise GameError
2.   return GameError

**Review by teds-lin — COMMENTED** [review]
LGTM

File: backend/app/domain/game.py (hunk 1 above)
Comment (metalalive) [reply backend/app/domain/game.py:87]: For coding style and the error-handling issue , let's postpone the review here and move to [this discussion thread](https://github.com/Game-as-a-Service/Pandemic-Reign-of-Cthulhu/discussions/30) first , then move back later whenever the backend developers make the decision.

Comment (teds-lin) [reply backend/app/domain/game.py:87]: > For coding style and the error-handling issue , let's postpone the review here and move to [this discussion thread](https://github.com/Game-as-a-Service/Pandemic-Reign-of-Cthulhu/discussions/30) first , then move back later whenever the backend developers make the decision.

同意, 另外處理好了

File: backend/app/adapter/api.py [hunk 2]
```diff
@@ -48,6 +54,14 @@ async def read_unselected_investigators(game_id: str):
     return result
 
 
+@_router.patch("/games/{game_id}/investigator", status_code=200)
+async def swtich_investigator(game_id: str, req: UpdateInvestigatorDto):
+    uc = SwitchInvestigatorUseCase(shared_context["repository"])
+    error = await uc.execute(game_id, req.player_id, req.investigator)
+    if error:
+        return JSONResponse(status_code=400, content={"reason": error.args[0]})
```
[the remark this answers is not in the document]
Comment (metalalive) [reply backend/app/adapter/api.py:62]: My bad, that is the implementation specific ONLY to [CPython](https://github.com/python/cpython/blob/main/Objects/exceptions.c) .
What about adding a data member `reason`  to the `GameError` ?

For example :
```python
class GameError(Exception):
    def __init__(self, r:str):
        self._reason = r

    @property
    def reason(self) -> str:
        return  self._reason
```

[the remark this answers is not in the document]
Comment (teds-lin) [reply backend/app/adapter/api.py:62]: 經過[這邊](https://github.com/Game-as-a-Service/Pandemic-Reign-of-Cthulhu/discussions/30#discussioncomment-7500088)討論後(usecase那層是raise exception)，可能會有類似下面code
```python
from fastapi import HTTPException

from app.domain.game import GameError

class GameNotFoundError(HTTPException):
    pass

class InvestigatorError(HTTPException):
    pass

@_router.patch("/games/{game_id}/investigator") 
async def switch_investigator(game_id, req):
    try:
        uc = SwitchInvestigatorUseCase()
        error = await uc.execute(game_id, req.player_id, req.investigator)
        if error:
            if error.args[0] == "invalid-game": 
                raise GameNotFoundError()
            else:
                raise InvestigatorError()
    except GameError:
        raise GameNotFoundError()
```

[the remark this answers is not in the document]
Comment (metalalive) [reply backend/app/adapter/api.py:62]: Absolutely,
different error types should be declared to different exception classes, such as `GameNotFoundError`, `InvestigatorError`.

[the remark this answers is not in the document]
Comment (metalalive) [reply backend/app/adapter/api.py:62]: In the domain model / entity layer, I am considering to :
- rename `GameError` to `BaseGameError`
  - remove the data member : `reason`
- extend `BaseGameError` to other different types of errors which would happen in the `Game` class, such as :
  - `IncorrectNumPlayersError`
  - `InvalidInvestigatorError`
  - `InvestigatorAlreadyChosenError`
  - `InvalidPlayerError`
  - `GameNotFoundError`

Each error above corresponds to another HTTP exception declared in the API endpoint , with similar naming convention

| domain layer | API endpoint |  HTTP status code | 
|-----|-----|-----|
| `IncorrectNumPlayersError` | `IncorrectNumPlayersHTTPException` | 400 |
| `InvalidInvestigatorError` | `InvalidInvestigatorHTTPException` |  400 |
|`InvestigatorAlreadyChosenError` | `InvestigatorAlreadyChosenHTTPException` |  409, Confict |
| `InvalidPlayerError` | `InvalidPlayerHTTPException` |  400 |
| `GameNotFoundError` | `GameNotFoundHTTPException` |  404, Not Found |

File: backend/app/domain/game.py (hunk 1 above)
Comment (metalalive) [reply backend/app/domain/game.py:87]: I will modify the following part of code
- change `return GameError` to `raise GameError` in :
  - `app/domain/game.py`
  - `app/usecase/config_game.py`
- `Game.switch_character()` catches `GameError` for state rollback, then raise the same error again to upper-level code
  ```python
    def switch_character(
        self, player_id: str, new_invstg: Investigator
    ):
        player = self.get_player(player_id)
        if player is None:
            raise GameError("invalid-player")
        old_invstg = player.get_investigator()
        try:
            if old_invstg :
                self._clear_character_selection(old_invstg)
            self.assign_character(new_invstg)
            player.set_investigator(new_invstg)
        except GameError as e : # roll back to previous state
            if old_invstg:
                self.assign_character(old_invstg)
            raise

  ```

File: backend/app/adapter/api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (teds-lin) [reply backend/app/adapter/api.py:62]: 這樣要注意當後續實作業務邏輯變得多且複雜時，相關的錯誤類別對應的subclass會多到導致維護困難

**如果**真的想要細分的話，可考慮使用Error code 
(`InvalidPlayerError`的HTTP狀態碼我改成`422`)
```python
from fastapi import HTTPException
from enum import Enum

class GameErrorCodes(Enum):
  INCORECT_NUM_PLAYERS = (1001, 400)  
  INVALID_INVESTIGATOR = (1002, 400)
  INVESTIGATOR_CHOSEN = (1003, 409)
  INVALID_PLAYER = (1004, 422)  
  GAME_NOT_FOUND = (1005, 404)

class GameError(Exception):
  def __init__(self, code: GameErrorCodes, message: str = None): 
    self.code = code
    self.message = message


class GameHTTPException(HTTPException):
  def __init__(self, err: GameError):
    status_code = err.code.value[1]
    super().__init__(status_code=status_code, 
                     detail=err.message or err.code.name)


try:
  # game logic
except GameError as e:
  raise GameHTTPException(e)
``` 

延伸有關於error code的[討論](https://discord.com/channels/937992003415838761/1098892690562752512/1098904395925688351)

[the remark this answers is not in the document]
Comment (metalalive) [reply backend/app/adapter/api.py:62]:
```python
class GameErrorCodes(Enum):
  INCORECT_NUM_PLAYERS = (1001, 400)  
  INVALID_INVESTIGATOR = (1002, 400)
  INVESTIGATOR_CHOSEN = (1003, 409)
  INVALID_PLAYER = (1004, 422)  
  GAME_NOT_FOUND = (1005, 404)
```
This is also OK to me, I've been reading some 3rd-party Python packages , the way the code maintainers organize all errors could be :
- subclass the base error class
- setup `code` or  `type` data member  to the error class

[the remark this answers is not in the document]
Comment (metalalive) [reply backend/app/adapter/api.py:62]: I edited my comment above several times  , please refresh the page before you read it , thanks

[the remark this answers is not in the document]
Comment (metalalive) [reply backend/app/adapter/api.py:62]: I edited my comment above several times , please refresh the page before you read it , thanks

[the remark this answers is not in the document]
Comment (teds-lin) [reply backend/app/adapter/api.py:62]: >What if there are 2 different functions raising the errors with the same error code ?

粗略想到有幾種作法：

-  就像discord [討論串](https://discord.com/channels/937992003415838761/1098892690562752512/1098965777085894766)提到的一樣，雖然都是`好友已滿`，可是依照引發的錯誤function不同，還是會編成不同的error code
   - 錯誤碼含function name
    ASSIGN_CHARACTER_INVALID_INVESTIGATOR = 101  
    SWITCH_CHARACTER_INVALID_INVESTIGATOR = 102
   - 將錯誤碼進行分組(group)
     code 1xx就是assign_character引發錯誤
     code 2xx就是switch_character引發錯誤
-  `GameError` class增加function_name : `GameError.function_name ="assign_character"`

```python
class GameError(Exception):
  def __init__(self, code: GameErrorCodes, message: str = None, function_name: str = None): 
    self.code = code
    self.message = message
    self.function_name = function_name

class GameHTTPException(HTTPException):
  def __init__(self, err: GameError):
    status_code = err.code.value[1]
    super().__init__(status_code=status_code, 
                     detail=f"{err.function_name}: {err.message or err.code.name}")

def assign_character():
  # game logic
  raise GameError(GameErrorCodes.INVALID_PLAYER, "invalid-player", "assign_character")

def switch_character():
  # game logic
  raise GameError(GameErrorCodes.INVALID_PLAYER, "invalid -layer", "switch_character")
```

[the remark this answers is not in the document]
Comment (metalalive) [reply backend/app/adapter/api.py:62]: From the [Discord discussion](https://discord.com/channels/937992003415838761/1098892690562752512/1098904395925688351) :

> custom code可能就6碼來查表
前3碼代表對應function, 例如傳送好友邀請，後3碼為具體錯誤樣態

Here is possible implementaion for backend function code
```python
class GameFuncCodes(Enum):
  ADD_PLAYERS = 1001  
  ASSIGN_CHARACTER  = 1002
  CLEAR_CHARACTER_SELECTION  = 1003
```

The `GameError` can be modified to 
```python
class GameError(Exception):
  def __init__(self, ecode: GameErrorCodes , fncode:GameFuncCodes, message: str = None): 
    self.ecode = ecode
    self.fncode = fncode
    self.message = message
```

[the remark this answers is not in the document]
Comment (metalalive) [reply backend/app/adapter/api.py:62]: I edited my comment above several times , please refresh the page before you read it , thanks

[the remark this answers is not in the document]
Comment (metalalive) [reply backend/app/adapter/api.py:62]: OK , I'll finalize the discussion here and start implementing relevant error classes,
for any other issue , please start another GitHub Issue, thanks

[the remark this answers is not in the document]
Comment (metalalive) [reply backend/app/adapter/api.py:62]: Remind that If the backend throws  `fastapi.HTTPException`  in the API endpoints , the response body will look like :
```json
{"detail": {"reason": 1234} }
``` 

Previously I used `fastapi.responses.JSONResponse` , it allows to customize response body so it matches the [API specification](https://game-as-a-service.github.io/Pandemic-Reign-of-Cthulhu/doc/api/) of the game : 
```json
{"reason": 1234}
```
which approaches do you prefer ?

File: backend/app/domain/game.py [hunk 3]
```diff
@@ -65,15 +95,49 @@ def match_chk(p):
...
+        # TODO, figure out better design options
+        if investigator not in self._investigators:
+            raise GameError(
+                e_code=GameErrorCodes.INVALID_INVESTIGATOR,
+                fn_code=GameFuncCodes.CLEAR_CHARACTER_SELECTION,
+            )
+
+        self._investigators[investigator] = False
+
+    def switch_character(self, player_id: str, new_invstg: Investigator):
```
Comment (teds-lin) [inline backend/app/domain/game.py:123]: 將`switch_character`業務邏輯拆更細， @metalalive  覺得這樣如何？

```python
    def _assign_new_character(self, player: Player, new_invstg: Investigator):
        self._validate_investigator(new_invstg)
        self.assign_character(new_invstg)
        player.set_investigator(new_invstg)

    def _validate_investigator(self, investigator: Investigator):
        if investigator not in self._investigators:
            raise GameError(
                e_code=GameErrorCodes.INVALID_INVESTIGATOR,
                fn_code=GameFuncCodes.ASSIGN_CHARACTER,
            )

    def switch_character(self, player_id: str, new_invstg: Investigator):
        player = self.get_player(player_id)
        if player is None:
            raise GameError(
                e_code=GameErrorCodes.INVALID_PLAYER,
                fn_code=GameFuncCodes.SWITCH_CHARACTER,
            )
        old_invstg = player.get_investigator()
        try:
            if old_invstg:
                self._clear_character_selection(old_invstg)
            self._assign_new_character(player, new_invstg)
        except GameError as e:
            if old_invstg and e.func_code == GameFuncCodes.ASSIGN_CHARACTER:
                # roll back to previous state
                self._assign_new_character(player, old_invstg)
            raise
```

**Review by teds-lin — APPROVED** [review]
LGTM

@metalalive 給你決定 `switch_character` 是否要拆分更細緻的業務邏輯?
我對是否修改沒意見

File: backend/app/domain/game.py (hunk 3 above)
Comment (metalalive) [reply backend/app/domain/game.py:123]: Allow me to postpone the code change in the future when this is necessary. Here is my thought :

- It looks like the validation part can be reused by `_clear_character_selection()`  and `assign_character()` , however the code is reduced by just 4 lines after the modification.
- It is unlikely to add more logic and code to the functions `switch_character()` and  `assign_character()` .
  - Remind that  these functions are invoked only before the game starts and the players have not determined which character to choose
  - After the game starts, the 2 functions should not be invoked as the players are stick to their chosen characters.

**Pull request merged**